### ISSUE TITLE
Error: Cannot write property

### ISSUE THREAD

**Issue by rowhel95** [opis/closure#28]
I have a class that extends mysqli class and those properties are defined as public but are actually read only properties so I have to put $property->setValue() and $property->getValue() inside a try and catch block for it to work.

Example:
```
do{
    foreach ($reflection->getProperties() as $property){
        if($property->isStatic()){
            continue;
        }
        $property->setAccessible(true);
        $value = $property->getValue($instance);
        if(is_array($value) || is_object($value)){
            $this->mapByReference($value);
        }

        try {
            $property->setValue($data, $value);
        }
        catch(\Exception $e) {}
        catch(\Throwable $e) {}
    }
} while($reflection = $reflection->getParentClass());
```
and
```
do{
    if(!$reflection->isUserDefined()){
        break;
    }
    foreach ($reflection->getProperties() as $property){
        if($property->isStatic()){
            continue;
        }
        $property->setAccessible(true);

        try {
            $item = $property->getValue($data);
        } catch(\Exception $e) {
            continue;
        } catch(\Throwable $e) {
            continue;
        }

        if ($item instanceof SerializableClosure || ($item instanceof SelfReference && $item->hash === $this->code['self'])) {
            $this->code['objects'][] = array(
                'instance' => $data,
                'property' => $property,
                'object' => $item instanceof SelfReference ? $this : $item,
            );
        } elseif (is_array($item) || is_object($item)) {
            $this->mapPointers($item);
            $property->setValue($data, $item);
        }
    }
} while($reflection = $reflection->getParentClass());
```

**Comment by sorinsarca** [comment]
@rowhel95 please provide a code snippet so we can reproduce your problem

**Comment by rowhel95** [comment]
@sorinsarca [](https://bugs.php.net/bug.php?id=63591)

**Comment by sorinsarca** [comment]
That is not related to opis/closure, please provide a php code that reproduces your problem when you try to serialize/unserialize something using opis/closure.

**Comment by msarca** [comment]
You cannot serialize database connections in PHP. The `try-catch` thing will not fix this, it will only hide it. I'm closing this issue because it has nothing to do with opis/closure.

**Comment by rowhel95** [comment]
```
Error: Cannot write property in /vendor/opis/closure/src/SerializableClosure.php:625
Stack trace:
#0 /vendor/opis/closure/src/SerializableClosure.php(625): ReflectionProperty->setValue(Object(database), 1)
#1 /vendor/opis/closure/src/SerializableClosure.php(581): Opis\Closure\SerializableClosure->mapByReference(Object(database))
#2 /vendor/opis/closure/src/SerializableClosure.php(148): Opis\Closure\SerializableClosure->mapByReference(Array)
#3 [internal function]: Opis\Closure\SerializableClosure->serialize()
#4 /class.base_table.php(2650): serialize(Object(Table\Sql))
#5 /class.base_table.php(493): framework\base_table->save()
```

**Comment by sorinsarca** [comment]
@rowhel95 that is the stack trace only, I want to generate that trace on my computer (in other words I want to reproduce the problem), so I needed an example of PHP code.

Anyway, as @msarca said, what are you trying to do is not possible in PHP.

**Comment by rowhel95** [comment]
@msarca @sorinsarca 

create a class named for example database that extends mysqli 

`class database extends mysqli`

now run this code

```
$oDatabase = new database();

serialize($oDatabase);

$test = function() use ($oDatabase) {

};

$test = new \Opis\Closure\SerializableClosure($test);

serialize($test);
```
you will see that serializing with php works just fine but using it for closure gives exception

**Comment by msarca** [comment]
I was able to reproduce this and I will provide a fix.

**Comment by msarca** [comment]
This is fixed in [v3.1.6](https://github.com/opis/closure/releases/tag/3.1.6).
